File: structmem.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

class Program {
    static void Main() {
        var a = new Seq { Id = 5, Vals = new List<int> { 6, 7 } };
        ReportChanges(a);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void ReportChanges(Seq a) {
        var b = UppedId(a);
        Console.WriteLine(a.Id);
        Console.WriteLine(b.Id);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static Seq UppedId(Seq sequence) {
        sequence.Id += 1;
        return sequence;
    }
}

struct Seq {
    public int Id;
    public List<int> Vals;
}

File: structmem.cpp
#include <iostream>
#include <memory>
#include <vector>

auto print_int(int i) -> void {
    std::cout << i << "\n";
}

struct Seq {
    int id;
    std::shared_ptr<std::vector<int>> vals;
};

auto upped_id(Seq seq) -> Seq {
    seq.id += 1;
    return seq;
}

auto report_changes(Seq a) -> void {
    auto b = upped_id(a);
    print_int(a.id);
    print_int(b.id);
}

auto main() -> int {
    auto a = Seq {
        .id = 5,
        .vals = std::make_shared<std::vector<int>>(std::vector { 6, 7 }),
    };
    report_changes(a);
}
